File: lib/api/models/ai_prompt_template.dart
// lib/models/ai_prompt_template.dart

class AIPromptTemplate {
  final int? templateId;
  final String name;
  final String purpose;
  final String templateText;
  final Map<String, dynamic>? parameters;
  final int createdByUserId;
  final DateTime? createdAt;
  final DateTime? lastUpdated;
  final bool isActive;

  // Relationships
  final Map<String, dynamic>? user;
  final List<Map<String, dynamic>>? interactions;

  AIPromptTemplate({
    this.templateId,
    required this.name,
    required this.purpose,
    required this.templateText,
    this.parameters,
    required this.createdByUserId,
    this.createdAt,
    this.lastUpdated,
    this.isActive = true,
    this.user,
    this.interactions,
  });

  factory AIPromptTemplate.fromJson(Map<String, dynamic> json) {
    return AIPromptTemplate(
      templateId: json['template_id'],
      name: json['name'],
      purpose: json['purpose'],
      templateText: json['template_text'],
      parameters: json['parameters'] != null
          ? Map<String, dynamic>.from(json['parameters'])
          : null,
      createdByUserId: json['created_by_user_id'],
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      lastUpdated: json['last_updated'] != null
          ? DateTime.parse(json['last_updated'])
          : null,
      isActive: json['is_active'] ?? true,
      user:
          json['user'] != null ? Map<String, dynamic>.from(json['user']) : null,
      interactions: (json['interactions'] as List?)
          ?.map((e) => Map<String, dynamic>.from(e))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (templateId != null) 'template_id': templateId,
      'name': name,
      'purpose': purpose,
      'template_text': templateText,
      if (parameters != null) 'parameters': parameters,
      'created_by_user_id': createdByUserId,
      if (createdAt != null) 'created_at': createdAt!.toIso8601String(),
      if (lastUpdated != null) 'last_updated': lastUpdated!.toIso8601String(),
      'is_active': isActive,
      if (user != null) 'user': user,
      if (interactions != null) 'interactions': interactions,
    };
  }

  AIPromptTemplate copyWith({
    int? templateId,
    String? name,
    String? purpose,
    String? templateText,
    Map<String, dynamic>? parameters,
    int? createdByUserId,
    DateTime? createdAt,
    DateTime? lastUpdated,
    bool? isActive,
    Map<String, dynamic>? user,
    List<Map<String, dynamic>>? interactions,
  }) {
    return AIPromptTemplate(
      templateId: templateId ?? this.templateId,
      name: name ?? this.name,
      purpose: purpose ?? this.purpose,
      templateText: templateText ?? this.templateText,
      parameters: parameters ?? this.parameters,
      createdByUserId: createdByUserId ?? this.createdByUserId,
      createdAt: createdAt ?? this.createdAt,
      lastUpdated: lastUpdated ?? this.lastUpdated,
      isActive: isActive ?? this.isActive,
      user: user ?? this.user,
      interactions: interactions ?? this.interactions,
    );
  }
}

File: final_project/models/ai_prompt_template.py
from datetime import datetime
from enum import Enum

from sqlalchemy import Column, Integer, String, Text, Boolean, DateTime, CheckConstraint, ForeignKey, Index, event
from sqlalchemy.dialects.postgresql import JSON
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from final_project.db import Base
from final_project.models.user import UserDB

# Enum for prompt template purposes
# class PromptTemplatePurpose(str, Enum):
#     TASK_SUGGESTION = "task_suggestion"
#     HABIT_RECOMMENDATION = "habit_recommendation"
#     CHAT_RESPONSE = "chat_response"

class AIPromptTemplateDB(Base):
    __tablename__ = "ai_prompt_template"

    template_id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(100), nullable=False)
    purpose = Column(String(50), nullable=False)
    template_text = Column(Text, nullable=False)
    parameters = Column(JSON, nullable=True)

    # Reference to the user who created this template
    created_by_user_id = Column(Integer, ForeignKey("users.user_id"), nullable=False)

    created_at = Column(DateTime, server_default=func.now(), nullable=False)
    last_updated = Column(DateTime, server_default=func.now(), nullable=False)
    is_active = Column(Boolean, default=True, nullable=False)

    __table_args__ = (
        CheckConstraint(
            "purpose IN ('task_suggestion', 'habit_recommendation', 'chat_response')",
            name="check_ai_prompt_template_purpose"
        ),
        Index("idx_ai_prompt_template_purpose", "purpose"),
        Index("idx_ai_prompt_template_is_active", "is_active"),
    )

        # Relationship to UserDB
    user = relationship(
        UserDB,
        back_populates="ai_prompt_templates"
    )

    # Relationship to AI interactions
    interactions = relationship(
        "AIInteractionDB",
        back_populates="prompt_template",
        cascade="all, delete-orphan"
    )

# Event listener to update 'last_updated' timestamp before updates
@event.listens_for(AIPromptTemplateDB, 'before_update')
def _update_last_updated(mapper, connection, target):
    target.last_updated = datetime.utcnow()
